wikidata/search.py

```
@@ -0,0 +1,415 @@
+
+        # Reset buffers
+        self.name_buffer = []
+        self.id_buffer = []
+        self.description_buffer = []
+
+
+def _calculate_similarity(query: str, name: str) -> float:
+    """
+    Calculate similarity score between search query and entity name.
+    
+    Uses :class:`difflib.SequenceMatcher` to compute a ratio and converts to percentage.
+    
+    :param query: The search query string
+    :type query: :class:`str`
+    :param name: The entity name to compare against
+    :type name: :class:`str`
+    :return: Similarity score as a percentage (0-100)
+    :rtype: :class:`float`
+    """
+    ratio = SequenceMatcher(None, query.lower(), name.lower()).ratio()
+    score = ratio * 100
+    return score
+
+
+def suggest_query(query: str, max_suggestions: int = 5) -> List[str]:
+    """
+    Suggest corrections for misspelled or ambiguous search queries.
+    
+    This function searches Wikidata and uses the top results to suggest
+    similar entity names that might be what the user intended to search for.
+    
+    :param query: The search query that may contain typos or be ambiguous
+    :type query: :class:`str`
+    :param max_suggestions: Maximum number of suggestions to return (default: 5)
+    :type max_suggestions: :class:`int`
+    :return: List of suggested query strings sorted by relevance
+    :rtype: :class:`List`\\ [:class:`str`]
+    
+    """
+    try:
+        # Search for the query to get similar entity names
+        results = _search(query, limit=50)
+        
+        if not results:
+            return []
+        
+        # Extract entity names from results
+        entity_names = [result.name for result in results]
+        
+        # Use difflib to find close matches
+        suggestions = get_close_matches(
+            query,
+            entity_names,
+            n=max_suggestions,
+            cutoff=0.6
+        )
+        
+        return suggestions
+        
+    except Exception as e:
+        return []
```

Catching all exceptions and returning an empty list silently hides errors. Consider either logging the exception or letting it propagate.